### PR TITLE
Add change delegate ruleset with refactor

### PR DESCRIPTION
**Pull request by KrickRay**
- Смена рулсета делегата
- Клонирование рулсета делегата
- Небольшой рефактор
- Фиксы

### CODE REVIEW

File: src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                filter((r) => r === 'confirm'),
+                switchMap(() =>
+                    this.routingRulesService.deleteDelegate({
+                        mainRulesetRefID: delegateId.parentRefId,
+                        delegateIdx: delegateId.delegateIdx,
+                    })
+                ),
+                untilDestroyed(this)
+            )
+            .subscribe({ error: this.errorService.error });
```
Comment (KisyaLisya) [inline src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.ts:137]: unreadable. It works but really unreadable

File: src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        fixture.detectChanges();
+    });
+
+    it('should be created', () => {
+        expect(component).toBeTruthy();
+    });
+
+    describe('methods', () => {});
+
+    describe('template', () => {});
```
Comment (KisyaLisya) [inline src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.spec.ts:37]: same question as before

File: src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { ChangeDetectionStrategy, Component } from '@angular/core';
+
+@Component({
+    selector: 'cc-routing-rules-select',
+    templateUrl: 'routing-rules-select.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class RoutingRulesSelectComponent {}
```
Comment (KisyaLisya) [inline src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.ts:8]: Component for what? It's a ng-content only

File: src/app/thrift-services/damsel/domain-cache.service.ts [hunk 4]
```diff
@@ -61,17 +60,26 @@ export class DomainCacheService {
             )
         );
 
-    commit = (commit: Commit, version?: Version) => {
-        return (version ? of(version) : this.version$).pipe(
-            take(1),
+    commit = (commit: Commit, version?: Version | number, reload = true) => {
+        const version$ = version ? of(version) : this.version$.pipe(take(1));
+        return version$.pipe(
             switchMap((v) =>
```
Comment (KisyaLisya) [inline src/app/thrift-services/damsel/domain-cache.service.ts:66]: can we name variables not in one letter?

File: src/app/thrift-services/damsel/routing-rules/utils/get-delegate.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { RoutingDelegate, RoutingRulesObject } from '../../gen-model/domain';
+
+export function getDelegate(ruleset: RoutingRulesObject, delegateIdx: number): RoutingDelegate {
```
Comment (KisyaLisya) [inline src/app/thrift-services/damsel/routing-rules/utils/get-delegate.ts:3]: it returns undefined too

File: src/app/thrift-services/utils/create-next-id.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export function createNextId(ids: number[]) {
+    return Math.max(...ids) + 1;
```
Comment (KisyaLisya) [inline src/app/thrift-services/utils/create-next-id.ts:2]: Do you handling numeric overflow somehow?

File: src/components/confirm-action-dialog/confirm-action-dialog.component.html [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<h2 class="mat-headline">{{ data.title ? data.title : 'Confirm this action' }}</h2>
+<h2 class="mat-headline">{{ data?.title || 'Confirm this action' }}</h2>
```
Comment (KisyaLisya) [inline src/components/confirm-action-dialog/confirm-action-dialog.component.html:1]: ```data.title ?? 'Confrim ...'```

**Review by KisyaLisya — COMMENTED** [review]
Very strange

File: src/app/thrift-services/utils/create-next-id.ts (hunk 6 above)
Comment (KrickRay) [reply src/app/thrift-services/utils/create-next-id.ts:2]: we have Int64, but it will also be possible to use deleted ids

Comment (KrickRay) [reply src/app/thrift-services/utils/create-next-id.ts:2]: but yes, here we will rest on MAX_SAFE_INTEGER

File: src/components/confirm-action-dialog/confirm-action-dialog.component.html (hunk 7 above)
Comment (KrickRay) [reply src/components/confirm-action-dialog/confirm-action-dialog.component.html:1]: angular not support

File: src/app/thrift-services/damsel/domain-cache.service.ts (hunk 4 above)
Comment (KrickRay) [reply src/app/thrift-services/damsel/domain-cache.service.ts:66]: here is a short context, we are practicing this code

File: src/app/thrift-services/damsel/routing-rules/utils/get-delegate.ts (hunk 5 above)
Comment (KrickRay) [reply src/app/thrift-services/damsel/routing-rules/utils/get-delegate.ts:3]: do you still believe in strict mode?

File: src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.ts (hunk 1 above)
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.ts:137]: why?

File: src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.ts (hunk 3 above)
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/routing-rules-select/routing-rules-select.component.ts:8]: del

File: src/app/sections/payment-routing-rules/change-delegate-ruleset-dialog/change-delegate-ruleset-dialog.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        fixture.detectChanges();
+    });
+
+    it('should be created', () => {
+        expect(component).toBeTruthy();
+    });
+
+    describe('methods', () => {});
+
+    describe('template', () => {});
```
[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/change-delegate-ruleset-dialog/change-delegate-ruleset-dialog.component.spec.ts]: just generated

[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/change-delegate-ruleset-dialog/change-delegate-ruleset-dialog.component.spec.ts]: del

File: src/app/sections/payment-routing-rules/change-delegate-ruleset-dialog/change-delegate-ruleset-dialog.component.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import { RoutingRulesService } from '../../../thrift-services';
+
+@UntilDestroy()
+@Component({
+    selector: 'cc-change-delegate-ruleset-dialog',
+    templateUrl: 'change-delegate-ruleset-dialog.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ChangeDelegateRulesetDialogComponent implements OnInit {
+    static defaultConfig: MatDialogConfig = {
```
[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/change-delegate-ruleset-dialog/change-delegate-ruleset-dialog.component.ts:16]: i will change

File: src/app/sections/payment-routing-rules/party-delegate-rulesets/party-delegate-rulesets.component.ts [hunk 10]
```diff
@@ -25,28 +20,38 @@ import { PartyDelegateRulesetsService } from './party-delegate-rulesets.service'
...
+                    text: r?.paymentInstitution?.data?.name,
+                    caption: r?.paymentInstitution?.ref?.id,
+                },
+                mainRuleset: {
+                    text: r?.mainRoutingRule?.data?.name,
+                    caption: r?.mainRoutingRule?.ref?.id,
+                },
+                partyDelegate: {
+                    text: r?.partyDelegate?.description,
+                    caption: r?.partyDelegate?.ruleset?.id,
```
[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/party-delegate-rulesets/party-delegate-rulesets.component.ts:46]: I will think about where it is necessary where not

File: src/app/sections/payment-routing-rules/party-payment-routing-ruleset/party-payment-routing-ruleset.component.ts [hunk 11]
```diff
@@ -30,44 +22,41 @@ const DIALOG_WIDTH = '548px';
...
-            return data;
-        }),
+        map(([ruleset, shops]) =>
+            ruleset.data.decisions.delegates
+                .filter((d) => d?.allowed?.condition?.party?.definition?.shop_is)
+                .map((d) => {
+                    const shopId = d.allowed.condition.party.definition.shop_is;
+                    return {
+                        parentRefId: ruleset.ref.id,
+                        delegateIdx: ruleset.data.decisions.delegates.findIndex((dt) => dt === d),
```
[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/party-payment-routing-ruleset/party-payment-routing-ruleset.component.ts:40]: i will rename

File: src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.html [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                        </button>
+                        <button mat-menu-item (click)="delete(element)">
+                            Delete
+                        </button>
+                    </mat-menu>
+                </td>
+            </ng-container>
+            <tr
+                mat-header-row
+                *matHeaderRowDef="displayedColumns?.concat(endColumns)?.map(getColumnsKeys)"
```
[the remark this answers is not in the document]
Comment (KrickRay) [reply src/app/sections/payment-routing-rules/routing-rules-list/routing-rules-list.component.html:59]: but not undefined.concat

File: src/app/thrift-services/utils/create-next-id.ts (hunk 6 above)
Comment (KrickRay) [reply src/app/thrift-services/utils/create-next-id.ts:2]: added todo

**Review by KisyaLisya — APPROVED** [review]

**Review by KisyaLisya — APPROVED** [review]

**Review by KeinAsylum — APPROVED** [review]

**Pull request merged**